src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/FlagAuthoritySystem.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.ligthandshadow.componentsystem.controllers;
+
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.EventPriority;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.ligthandshadow.componentsystem.LASUtils;
+import org.terasology.logic.delay.DelayManager;
+import org.terasology.logic.delay.DelayedActionTriggeredEvent;
+import org.terasology.logic.inventory.ItemComponent;
+import org.terasology.logic.inventory.events.DropItemEvent;
+import org.terasology.logic.inventory.events.GiveItemEvent;
+import org.terasology.registry.In;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.BlockManager;
+import org.terasology.world.block.items.BlockItemComponent;
+
+/**
+ * Handles events related to flag drops and pickups.
+ */
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class FlagAuthoritySystem extends BaseComponentSystem {
+
+    @In
+    DelayManager delayManager;
+
+    @In
+    WorldProvider worldProvider;
+
+    @In
+    BlockManager blockManager;
+
+    /**
+     * Add a delayed action using delay manager to flags when they are dropped.
+     * @see DelayManager
+     *
+     * @param event
+     * @param itemEntity
+     * @param itemComponent
+     * @param blockItemComponent
+     */
+    @ReceiveEvent
+    public void onDropItemEvent(DropItemEvent event, EntityRef itemEntity, ItemComponent itemComponent,
```

Comment:
Correction: Since FlagComponent would not be used further in the event handler, just having it specified in the annotation is enough, as in-
`@ReceiveEvent(components = ...`

Reply:
That sounds better… @darshan3 any issue with having `FlagComponent` as predicate here?

Reply:
The flags don't use the `FlagComponent`. For some reason, there are two different components `BlackFlagComponent` and `RedFlagComponent` which are added to the prefabs of the respective blocks. Hence, I wasn't able to implement it based on FlagComponent.

Reply:
In that case, can `FlagComponent` be removed? There's also `HasFlagComponent`, which seems somewhat the same... is that some legace component, or does it denote that the player is currently carrying the flag? In that case, rename it to `FlagCarrier` component?